File: apple/Packages/DsmNetwork/Tests/ContainerImagePullTests.swift
import DsmCore
import Foundation
import XCTest
@testable import DsmNetwork

final class ContainerImagePullTests: XCTestCase {
    func test启动与状态固定V1且保留字符串或整数任务编号() async throws {
        for format in [DsmRequestFormat.form, .json] {
            for numeric in [false, true] {
                let transport = MockHTTPTransport(responses: [reply(tags), reply(images),
                    reply("{\"success\":true,\"data\":{\"task_id\":\(numeric ? "42" : "\"synthetic-task\"")}}"),
                    reply(status()), reply(status(finished: true)), reply(images)])
                let repository = try makeRepository(transport, format: format)
                let request = makeRequest()
                let first = try await repository.startContainerImagePull(request)
                XCTAssertEqual(first.stage, .downloading); XCTAssertEqual(first.percentage, 25)
                let final = try await repository.reviewContainerImagePull(id: request.id)
                XCTAssertEqual(final?.stage, .ready)
                let repeated = try await repository.startContainerImagePull(request)
                XCTAssertEqual(repeated, final)
                let calls = await transport.recordedRequests()
                let start = try XCTUnwrap(calls.first { parameter("method", $0) == "pull_start" })
                let encodedRepository = try XCTUnwrap(parameter("repository", start))
                let decodedRepository = format == .json
                    ? try JSONDecoder().decode(String.self, from: Data(encodedRepository.utf8))
                    : encodedRepository
                XCTAssertEqual(decodedRepository, "synthetic/web")
                XCTAssertEqual(parameter("tag", start), format == .json ? "\"stable\"" : "stable")
                let statusRequest = try XCTUnwrap(calls.first { parameter("method", $0) == "pull_status" })
                XCTAssertEqual(parameter("task_id", statusRequest), numeric ? "42" : format == .json ? "\"synthetic-task\"" : "synthetic-task")
                XCTAssertTrue(calls.allSatisfy { parameter("version", $0) == "1" })
                XCTAssertEqual(calls.filter { parameter("method", $0) == "pull_start" }.count, 1)
            }
        }
    }

    func test旧标签与100百分比不能确认未结束任务() async throws {
        let transport = MockHTTPTransport(responses: [reply(tags), reply(images), reply(receipt), reply(status(current: 100))])
        let repository = try makeRepository(transport)
        let result = try await repository.startContainerImagePull(makeRequest())
        XCTAssertEqual(result.stage, .downloading); XCTAssertEqual(result.percentage, 100)
        XCTAssertEqual(result.outcome.status, .submittedButUnverified)
    }

    func test无回执和不安全数字任务编号不得猜测或重发() async throws {
        for taskID in ["null", "true", "-1", "9007199254740993"] {
            let transport = MockHTTPTransport(responses: [reply(tags), reply(images), reply("{\"success\":true,\"data\":{\"task_id\":\(taskID)}}")])
            let repository = try makeRepository(transport); let original = makeRequest()
            let result = try await repository.startContainerImagePull(original)
            let again = try await repository.startContainerImagePull(original)
            let conflicting = try await repository.startContainerImagePull(makeRequest())
            XCTAssertEqual(result.stage, .awaitingReceipt); XCTAssertEqual(again.stage, .awaitingReceipt)
            XCTAssertFalse(conflicting.outcome.submitted); XCTAssertEqual(conflicting.outcome.errorCategory, .conflict)
            let calls = await transport.recordedRequests(); XCTAssertEqual(calls.count, 3)
        }
    }

    func test回执丢失保存未知记录且独立核查不重放() async throws {
        let transport = MockHTTPTransport(steps: [.response(reply(tags)), .response(reply(images)), .urlError(.timedOut)])
        let repository = try makeRepository(transport); let request = makeRequest()
        let result = try await repository.startContainerImagePull(request)
        let pending = try await repository.loadContainerImagePulls()
        let reviewed = try await repository.reviewContainerImagePull(id: request.id)
        XCTAssertEqual(result.stage, .awaitingReceipt); XCTAssertEqual(pending.count, 1); XCTAssertEqual(reviewed?.stage, .awaitingReceipt)
        let calls = await transport.recordedRequests(); XCTAssertEqual(calls.count, 3)
    }

    func test完成必须有原生结束标记回显绑定和标签回读() async throws {
        for scenario in ["flag", "name", "tag", "missing-image", "partial-list"] {
            var taskStatus = status(finished: true)
            if scenario == "flag" { taskStatus = taskStatus.replacingOccurrences(of: "\"finished\":true", with: "\"finished\":\"true\"") }
            if scenario == "name" { taskStatus = taskStatus.replacingOccurrences(of: "docker.io/synthetic/web", with: "other/web") }
            if scenario == "tag" { taskStatus = taskStatus.replacingOccurrences(of: "stable", with: "latest") }
            let finalImages = scenario == "missing-image" ? "{\"success\":true,\"data\":{\"images\":[]}}" :
                scenario == "partial-list" ? images.replacingOccurrences(of: "\"total\":1", with: "\"total\":2") : images
            let transport = MockHTTPTransport(responses: [reply(tags), reply(images), reply(receipt), reply(taskStatus), reply(finalImages)])
            let repository = try makeRepository(transport)
            let result = try await repository.startContainerImagePull(makeRequest())
            XCTAssertEqual(result.stage, .needsReview)
            XCTAssertNotEqual(result.outcome.status, .confirmedSuccess)
        }
    }

    func test明确权限拒绝缓存为终态且不会回读成成功() async throws {
        let transport = MockHTTPTransport(responses: [reply(tags), reply(images), reply("{\"success\":false,\"error\":{\"code\":105}}")])
        let repository = try makeRepository(transport); let request = makeRequest()
        let result = try await repository.startContainerImagePull(request)
        let repeated = try await repository.startContainerImagePull(request)
        XCTAssertEqual(result.stage, .rejected); XCTAssertEqual(result.outcome.status, .permissionDenied); XCTAssertEqual(result, repeated)
        let pending = try await repository.loadContainerImagePulls(); XCTAssertTrue(pending.isEmpty)
        let calls = await transport.recordedRequests(); XCTAssertEqual(calls.count, 3)
    }

    func test缺确认或无效目标不得读取或启动() async throws {
        let transport = MockHTTPTransport(responses: [])
        let repository = try makeRepository(transport)
        for request in [ContainerImagePullRequest(repository: "synthetic/web", tag: "stable"),
                        ContainerImagePullRequest(repository: "user@registry.invalid", tag: "stable", isConfirmed: true)] {
            let result = try await repository.startContainerImagePull(request)
            XCTAssertFalse(result.outcome.submitted)
        }
        let calls = await transport.recordedRequests(); XCTAssertTrue(calls.isEmpty)
    }

    func test查询失败后重试只核查同一任务() async throws {
        let transport = MockHTTPTransport(steps: [.response(reply(tags)), .response(reply(images)), .response(reply(receipt)), .urlError(.networkConnectionLost),
            .response(reply(status(finished: true))), .response(reply(images))])
        let repository = try makeRepository(transport); let request = makeRequest()
        let first = try await repository.startContainerImagePull(request)
        let last = try await repository.reviewContainerImagePull(id: request.id)
        XCTAssertEqual(first.stage, .needsReview); XCTAssertEqual(last?.stage, .ready)
        let calls = await transport.recordedRequests(); XCTAssertEqual(calls.filter { parameter("method", $0) == "pull_start" }.count, 1)
    }

    func test启动后取消保留未知任务并且不发送取消接口() async throws {
        let transport = MockHTTPTransport(steps: [.response(reply(tags)), .response(reply(images)), .waitUntilCancelled])
        let repository = try makeRepository(transport); let request = makeRequest()
        let task = Task { try await repository.startContainerImagePull(request) }
        let deadline = Date().addingTimeInterval(2)
        while await transport.recordedRequests().count < 3 && Date() < deadline { await Task.yield() }
        task.cancel(); let result = try await task.value
        XCTAssertEqual(result.outcome.status, .cancellationRequestedAfterSubmission)
        let pending = try await repository.loadContainerImagePulls(); XCTAssertEqual(pending.count, 1)
        let calls = await transport.recordedRequests(); XCTAssertEqual(calls.count, 3)
    }

    func test预检读取时取消不启动也不登记未知任务() async throws {
        let transport = MockHTTPTransport(steps: [.waitUntilCancelled])
        let repository = try makeRepository(transport); let request = makeRequest()
        let task = Task { try await repository.startContainerImagePull(request) }
        let deadline = Date().addingTimeInterval(2)
        while await transport.recordedRequests().isEmpty && Date() < deadline { await Task.yield() }
        task.cancel(); let result = try await task.value
        XCTAssertEqual(result.outcome.status, .cancelledBeforeSubmission); XCTAssertFalse(result.outcome.submitted)
        let pending = try await repository.loadContainerImagePulls(); XCTAssertTrue(pending.isEmpty)
        let calls = await transport.recordedRequests(); XCTAssertEqual(calls.count, 1)
    }

    func test同标签下载与删除互斥() async throws {
        let transport = MockHTTPTransport(responses: [reply(tags), reply(images), reply(receipt), reply(status()), reply(images)])
        let repository = try makeRepository(transport)
        _ = try await repository.startContainerImagePull(makeRequest())
        let result = try await repository.deleteContainerImagesResult(ids: [ContainerImage.selectionID(imageID: "synthetic-id", repository: "synthetic/web", tag: "stable")])
        XCTAssertFalse(result.submitted); XCTAssertEqual(result.errorCategory, .conflict)
        let calls = await transport.recordedRequests(); XCTAssertFalse(calls.contains { parameter("method", $0) == "delete" })
    }

    private var tags: String { "{\"success\":true,\"data\":{\"tags\":[\"stable\"]}}" }
    private var images: String { "{\"success\":true,\"data\":{\"offset\":0,\"total\":1,\"images\":[{\"id\":\"synthetic-id\",\"repository\":\"synthetic/web\",\"tags\":[\"stable\"]}]}}" }
    private var receipt: String { "{\"success\":true,\"data\":{\"task_id\":\"synthetic-task\"}}" }
    private func status(finished: Bool = false, current: Int = 25) -> String {
        "{\"success\":true,\"data\":{\"finished\":\(finished),\"repository\":\"docker.io/synthetic/web\",\"tag\":\"stable\",\"current\":\(current),\"total\":100}}"
    }
    private func makeRequest() -> ContainerImagePullRequest { .init(repository: "synthetic/web", tag: "stable", isConfirmed: true) }
    private func reply(_ json: String) -> DsmHTTPResponse { .init(data: Data(json.utf8), statusCode: 200) }
    private func makeRepository(_ transport: any DsmHTTPTransport, format: DsmRequestFormat = .form) throws -> DsmServiceManagementRepository {
        let values = [DsmAPIName.dockerImage, DsmAPIName.dockerRegistry, DsmAPIName.dockerContainer].map { name in
            (name, ApiCapability(name: name, path: "entry.cgi", minVersion: 1, maxVersion: 2, requestFormat: format, selectedVersion: 2))
        }
        return try DsmServiceManagementRepository(profile: NasProfile(displayName: "Synthetic", host: "nas.example.invalid", port: 5001),
            capabilities: CapabilitySet(Dictionary(uniqueKeysWithValues: values)),
            session: AuthSession(sid: "SYNTHETIC_SESSION", synoToken: "SYNTHETIC_TOKEN", did: nil, isPortalPort: false), transport: transport)
    }
    private func parameter(_ name: String, _ request: URLRequest) -> String? {
        guard let body = request.httpBody, let text = String(data: body, encoding: .utf8) else { return nil }
        return URLComponents(string: "https://fixture.invalid/?" + text)?.queryItems?.first { $0.name == name }?.value
    }
}

File: windows/tests/LanStash.Tests/Containers/ContainerImagePullTests.cs
using System.Net;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using LanStash.Domain;
using LanStash.Infrastructure;

namespace LanStash.Tests.Containers;

public sealed class ContainerImagePullTests
{
    [Theory]
    [InlineData("FORM", false)]
    [InlineData("FORM", true)]
    [InlineData("JSON", false)]
    [InlineData("JSON", true)]
    public async Task StartAndStatusUseFixedVersionAndPreserveOpaqueTaskIdType(string format, bool numeric)
    {
        using var fixture = new Fixture(format) { NumericTaskId = numeric };
        var request = fixture.Request();
        Assert.True(fixture.Repository.CanPullImages);
        var first = await fixture.Repository.PullImageAsync(request);
        Assert.Equal(ContainerImagePullStage.Downloading, first.Stage); Assert.Equal(25, first.Percentage);
        var start = Assert.Single(fixture.Calls, call => call["method"] == "pull_start");
        Assert.Equal(format == "JSON" ? "\"synthetic/web\"" : "synthetic/web", start["repository"]);
        Assert.Equal(format == "JSON" ? "\"stable\"" : "stable", start["tag"]);
        Assert.Equal("1", start["version"]);
        AssertFixture(start, "pull-start", format);
        var status = Assert.Single(fixture.Calls, call => call["method"] == "pull_status");
        Assert.Equal(numeric ? "42" : format == "JSON" ? "\"synthetic-task\"" : "synthetic-task", status["task_id"]);
        if (!numeric) AssertFixture(status, "pull-status", format);
        Assert.DoesNotContain("password", start.Keys); Assert.DoesNotContain("registry", start.Keys);
        fixture.Finished = true;
        var ready = await fixture.Repository.ReviewImagePullAsync(request.RequestId);
        Assert.Equal(ContainerImagePullStage.Ready, ready!.Stage);
        Assert.Equal(MutationResultStatus.ConfirmedSuccess, ready.Outcome.Status);
        Assert.Empty(await fixture.Repository.GetImagePullRecoveriesAsync());
        Assert.Equal(ready, await fixture.Repository.PullImageAsync(request));
        Assert.Single(fixture.Calls, call => call["method"] == "pull_start");
    }

    [Fact]
    public async Task ExistingTagAndFullProgressCannotCompleteAnUnfinishedTask()
    {
        using var fixture = new Fixture("JSON") { Current = 100 };
        var result = await fixture.Repository.PullImageAsync(fixture.Request());
        Assert.Equal(ContainerImagePullStage.Downloading, result.Stage); Assert.Equal(100, result.Percentage);
        Assert.Equal(MutationResultStatus.SubmittedButUnverified, result.Outcome.Status);
        Assert.Single(await fixture.Repository.GetImagePullRecoveriesAsync());
    }

    [Theory]
    [InlineData("missing-image")]
    [InlineData("wrong-repository")]
    [InlineData("wrong-tag")]
    [InlineData("missing-finished")]
    [InlineData("string-finished")]
    [InlineData("partial-list")]
    [InlineData("status-rejected")]
    public async Task FinishedMustBeTypedBoundAndFollowedByCompleteImageReadback(string scenario)
    {
        using var fixture = new Fixture("JSON") { Finished = true, Scenario = scenario };
        var request = fixture.Request();
        var result = await fixture.Repository.PullImageAsync(request);
        Assert.Equal(ContainerImagePullStage.NeedsReview, result.Stage);
        Assert.NotEqual(MutationResultStatus.ConfirmedSuccess, result.Outcome.Status);
        fixture.Scenario = "";
        var final = await fixture.Reconnect().ReviewImagePullAsync(request.RequestId);
        Assert.Equal(ContainerImagePullStage.Ready, final!.Stage);
        Assert.Single(fixture.Calls, call => call["method"] == "pull_start");
    }

    [Theory]
    [InlineData(false)]
    [InlineData(true)]
    public async Task MissingReceiptNeverGuessesTaskOrReplaysByTarget(bool networkLoss)
    {
        using var fixture = new Fixture("FORM") { MissingReceipt = !networkLoss, LoseReceipt = networkLoss };
        var request = fixture.Request();
        Assert.Equal(ContainerImagePullStage.AwaitingReceipt, (await fixture.Repository.PullImageAsync(request)).Stage);
        Assert.Equal(ContainerImagePullStage.AwaitingReceipt, (await fixture.Reconnect().PullImageAsync(request)).Stage);
        var conflict = await fixture.Repository.PullImageAsync(request with { RequestId = Guid.NewGuid() });
        Assert.Equal(MutationErrorCategory.Conflict, conflict.Outcome.ErrorCategory);
        Assert.False(conflict.Outcome.Submitted);
        Assert.DoesNotContain(fixture.Calls, call => call["method"] == "pull_status");
        Assert.Single(fixture.Calls, call => call["method"] == "pull_start");
    }

    [Fact]
    public async Task ExplicitRejectionIsTerminalAndSameRequestCannotReplay()
    {
        using var fixture = new Fixture("JSON") { Reject = true };
        var request = fixture.Request(); var result = await fixture.Repository.PullImageAsync(request);
        Assert.Equal(ContainerImagePullStage.Rejected, result.Stage);
        Assert.Equal(MutationResultStatus.PermissionDenied, result.Outcome.Status);
        Assert.Equal(result, await fixture.Repository.PullImageAsync(request));
        Assert.Empty(await fixture.Repository.GetImagePullRecoveriesAsync());
        Assert.DoesNotContain(fixture.Calls, call => call["method"] == "pull_status");
    }

    [Theory]
    [InlineData("confirmation")]
    [InlineData("profile")]
    [InlineData("tag")]
    [InlineData("invalid-target")]
    public async Task InvalidPreflightDoesNotStart(string scenario)
    {
        using var fixture = new Fixture("JSON"); var request = fixture.Request();
        if (scenario == "confirmation") request = request with { RiskConfirmed = false };
        if (scenario == "profile") request = request with { ProfileId = Guid.NewGuid() };
        if (scenario == "tag") request = request with { Tag = "missing" };
        if (scenario == "invalid-target") request = request with { Repository = "user@registry.invalid" };
        Assert.False((await fixture.Repository.PullImageAsync(request)).Outcome.Submitted);
        Assert.DoesNotContain(fixture.Calls, call => call["method"] == "pull_start");
    }

    [Fact]
    public async Task CancellationAfterStartRetainsReceiptAndOnlyReviews()
    {
        using var cancellation = new CancellationTokenSource();
        using var fixture = new Fixture("JSON") { OnStart = cancellation.Cancel };
        var request = fixture.Request();
        var result = await fixture.Repository.PullImageAsync(request, cancellation.Token);
        Assert.Equal(MutationResultStatus.CancellationRequestedAfterSubmission, result.Outcome.Status);
        Assert.Single(await fixture.Repository.GetImagePullRecoveriesAsync());
        await fixture.Repository.ReviewImagePullAsync(request.RequestId);
        Assert.Single(fixture.Calls, call => call["method"] == "pull_start");
    }

    [Fact]
    public async Task PullAndDeleteOfSameTagAreMutuallyExclusive()
    {
        using var fixture = new Fixture("JSON");
        var request = fixture.Request();
        await fixture.Repository.PullImageAsync(request);
        var images = (await fixture.Repository.LoadSnapshotAsync()).Images.Items;
        var deletion = new ContainerImageDeleteRequest(fixture.Repository.ProfileId, images, Guid.NewGuid(), true);
        Assert.False((await fixture.Repository.DeleteImagesAsync(deletion)).Submitted);
        Assert.DoesNotContain(fixture.Calls, call => call["method"] == "delete");
    }

    [Fact]
    public async Task UnknownDeletionBlocksStartingPullOfSameTag()
    {
        using var fixture = new Fixture("FORM");
        var snapshot = await fixture.Repository.LoadSnapshotAsync();
        var deletion = await fixture.Repository.DeleteImagesAsync(new(fixture.Repository.ProfileId, snapshot.Images.Items, Guid.NewGuid(), true));
        Assert.Equal(MutationResultStatus.SubmittedButUnverified, deletion.Status);
        var pull = await fixture.Repository.PullImageAsync(fixture.Request());
        Assert.False(pull.Outcome.Submitted); Assert.Equal(MutationErrorCategory.Conflict, pull.Outcome.ErrorCategory);
        Assert.DoesNotContain(fixture.Calls, call => call["method"] == "pull_start");
    }

    [Fact]
    public async Task KnownTaskRemainsReviewableWithoutRegistryCapability()
    {
        using var fixture = new Fixture("JSON"); var request = fixture.Request();
        await fixture.Repository.PullImageAsync(request); fixture.Finished = true;
        var repository = fixture.Reconnect(includeRegistry: false);
        Assert.False(repository.CanPullImages);
        Assert.Equal(ContainerImagePullStage.Ready, (await repository.ReviewImagePullAsync(request.RequestId))!.Stage);
        Assert.Single(fixture.Calls, call => call["method"] == "pull_start");
    }

    [Theory]
    [InlineData(-1)]
    [InlineData(101)]
    public async Task InvalidProgressIsNotPresentedAsAValidPercentage(int current)
    {
        using var fixture = new Fixture("JSON") { Current = current };
        var result = await fixture.Repository.PullImageAsync(fixture.Request());
        Assert.Equal(ContainerImagePullStage.Downloading, result.Stage); Assert.Null(result.Percentage);
    }

    private static void AssertFixture(Dictionary<string, string> call, string operation, string format)
    {
        for (var directory = new DirectoryInfo(Directory.GetCurrentDirectory()); directory is not null; directory = directory.Parent)
        {
            var path = Path.Combine(directory.FullName, $"contracts/request-fixtures/container-manager/{operation}/synthetic-task/request.json");
            if (!File.Exists(path)) continue;
            var fixture = JsonNode.Parse(File.ReadAllText(path))!;
            Assert.Equal(fixture["api"]!["name"]!.GetValue<string>(), call["api"]);
            Assert.Equal(fixture["api"]!["method"]!.GetValue<string>(), call["method"]);
            foreach (var parameter in fixture["parameters"]!.AsArray())
            {
                var expected = parameter!["encodedValue"]!.GetValue<string>();
                Assert.Equal(format == "JSON" ? JsonSerializer.Serialize(expected) : expected, call[parameter["name"]!.GetValue<string>()]);
            }
            return;
        }
        throw new FileNotFoundException("缺少共享镜像下载夹具。");
    }

    private sealed class Fixture : IDisposable
    {
        private readonly NasProfile _profile = new(Guid.NewGuid(), "Synthetic", "nas.invalid", 5001, "synthetic-user");
        private readonly HttpClient _http; private readonly DsmApiClient _api; private readonly string _format;
        public IContainerManagerRepository Repository { get; }
        public bool NumericTaskId { get; init; }
        public bool MissingReceipt { get; init; }
        public bool LoseReceipt { get; init; }
        public bool Reject { get; init; }
        public bool Finished { get; set; }
        public int Current { get; init; } = 25;
        public string Scenario { get; set; } = "";
        public Action? OnStart { get; init; }
        public List<Dictionary<string, string>> Calls { get; } = [];
        public Fixture(string format) { _format = format; _http = new(new Handler(this)); _api = new(_http); Repository = Reconnect(); }
        public IContainerManagerRepository Reconnect(bool includeRegistry = true) => new DsmRepository(_profile, new(_profile.Id, "synthetic-sid", "synthetic-token", null), _api,
            new[] { "SYNO.Docker.Image", "SYNO.Docker.Registry", "SYNO.Docker.Container" }.Where(name => includeRegistry || name != "SYNO.Docker.Registry")
                .ToDictionary(name => name, name => new ApiCapability(name, "entry.cgi", 1, 2, _format)));
        public ContainerImagePullRequest Request() => new(_profile.Id, "synthetic/web", "stable", Guid.NewGuid(), true);
        private sealed class Handler(Fixture owner) : HttpMessageHandler
        {
            protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken token)
            {
                Assert.Equal(HttpMethod.Post, request.Method); Assert.Empty(request.RequestUri!.Query);
                var body = await request.Content!.ReadAsStringAsync(token);
                var form = body.Split('&').Select(part => part.Split('=', 2)).ToDictionary(part => WebUtility.UrlDecode(part[0]), part => WebUtility.UrlDecode(part[1]));
                owner.Calls.Add(form); var data = new JsonObject();
                if (form["api"] == "SYNO.Docker.Registry") data["tags"] = new JsonArray("stable", "latest");
                else if (form["api"] == "SYNO.Docker.Container") data["containers"] = new JsonArray();
                else if (form["method"] == "pull_start")
                {
                    owner.OnStart?.Invoke();
                    if (owner.Reject) return Reply("{\"success\":false,\"error\":{\"code\":105}}");
                    if (owner.LoseReceipt) throw new HttpRequestException("synthetic");
                    if (!owner.MissingReceipt) data["task_id"] = owner.NumericTaskId ? JsonValue.Create(42) : JsonValue.Create("synthetic-task");
                }
                else if (form["method"] == "pull_status")
                {
                    if (owner.Scenario == "status-rejected") return Reply("{\"success\":false,\"error\":{\"code\":400}}");
                    data["repository"] = owner.Scenario == "wrong-repository" ? "synthetic/other" : "docker.io/synthetic/web";
                    data["tag"] = owner.Scenario == "wrong-tag" ? "latest" : "stable";
                    if (owner.Scenario != "missing-finished") data["finished"] = owner.Scenario == "string-finished" ? JsonValue.Create("true") : JsonValue.Create(owner.Finished);
                    data["current"] = owner.Current; data["total"] = 100;
                }
                else if (form["method"] == "list")
                {
                    var afterStart = owner.Calls.Any(call => call["method"] == "pull_start");
                    var images = new JsonArray(new JsonObject { ["id"] = "synthetic-id", ["repository"] = "synthetic/web", ["tags"] = new JsonArray("stable") });
                    if (afterStart && owner.Scenario == "missing-image") images.Clear();
                    data["images"] = images; data["total"] = images.Count + (afterStart && owner.Scenario == "partial-list" ? 1 : 0); data["offset"] = 0;
                }
                else if (form["method"] == "delete") throw new HttpRequestException("合成删除回执丢失");
                else throw new InvalidOperationException("合成测试不允许其他写操作。");
                return Reply(new JsonObject { ["success"] = true, ["data"] = data }.ToJsonString());
            }
            private static HttpResponseMessage Reply(string json) => new(HttpStatusCode.OK) { Content = new StringContent(json, Encoding.UTF8, "application/json") };
        }
        public void Dispose() => _http.Dispose();
    }
}
